### ISSUE TITLE
gradient stats

### ISSUE THREAD

**Issue by a-maci** [ppwwyyxx/tensorpack#350]
I want to collect stats for gradients.

I put stat collection code at couple places (method 1 and method 2 below). The stats coming from these two methods are very different. **Any idea why and which one is correct?** If both methods are incorrect then any suggestions of where to put the stat collection code for gradients?

```
Method 1:
@tf.RegisterGradient("FGGrad")        #following the DoReFa style of code of monkey-patching
     def grad_fg(op, x):           # x is the grad tensor, DoReFa quantizes this but we compute stat and return
         add_activation_summary(x)
         if bitG == 32:
             return x
		 pass     
```                    
		
```
Method 2:
class SummaryGradient(MapGradient):       #code in tensorpack/tensorpack/tfutils/gradproc.py
    """
    Summary histogram, sparsity and RMS for each gradient variable.
    """

    def __init__(self):
        super(SummaryGradient, self).__init__(self._mapper)

    def _mapper(self, grad, var):
        name = var.op.name
        if name not in _summaried_gradient:
            _summaried_gradient.add(name)
            tf.summary.histogram(name + '-grad', grad)
            tf.summary.scalar(name + '-sparsity', tf.nn.zero_fraction(grad))
            add_moving_summary(rms(grad, name=name + '/rms'))
            add_moving_summary(tf.nn.zero_fraction(grad, name=name + '/sparsity'))
        return grad
```	
	
Results:	
```
 SummaryGradient/bn1/beta-sparsity: 0
 SummaryGradient/bn1/beta/sparsity/Mean: 0
 SummaryGradient/bn1/gamma-sparsity: 0
 SummaryGradient/bn1/gamma/sparsity/Mean: 0
 SummaryGradient/bn2/beta-sparsity: 0
 SummaryGradient/bn2/beta/sparsity/Mean: 0
 SummaryGradient/bn2/gamma-sparsity: 0
 SummaryGradient/bn2/gamma/sparsity/Mean: 0
 SummaryGradient/bn3/beta-sparsity: 0
 SummaryGradient/bn3/beta/sparsity/Mean: 0
 SummaryGradient/bn3/gamma-sparsity: 0
 SummaryGradient/bn3/gamma/sparsity/Mean: 0
 SummaryGradient/bn4/beta-sparsity: 0
 SummaryGradient/bn4/beta/sparsity/Mean: 0
 SummaryGradient/bn4/gamma-sparsity: 0
 SummaryGradient/bn4/gamma/sparsity/Mean: 0
 SummaryGradient/bn5/beta-sparsity: 0
 SummaryGradient/bn5/beta/sparsity/Mean: 0
 SummaryGradient/bn5/gamma-sparsity: 0
 SummaryGradient/bn5/gamma/sparsity/Mean: 0
 SummaryGradient/bn6/beta-sparsity: 0
 SummaryGradient/bn6/beta/sparsity/Mean: 0
 SummaryGradient/bn6/gamma-sparsity: 0
 SummaryGradient/bn6/gamma/sparsity/Mean: 2.1251e-07
 SummaryGradient/bn7/beta-sparsity: 0
 SummaryGradient/bn7/beta/sparsity/Mean: 0
 SummaryGradient/bn7/gamma-sparsity: 0
 SummaryGradient/bn7/gamma/sparsity/Mean: 0
 SummaryGradient/conv1/W-sparsity: 0
 SummaryGradient/conv1/W/sparsity/Mean: 0
 SummaryGradient/conv2/W-sparsity: 0
 SummaryGradient/conv2/W/sparsity/Mean: 1.0765e-07
 SummaryGradient/conv3/W-sparsity: 0
 SummaryGradient/conv3/W/sparsity/Mean: 0
 SummaryGradient/conv4/W-sparsity: 0
 SummaryGradient/conv4/W/sparsity/Mean: 2.4277e-09
 SummaryGradient/conv5/W-sparsity: 0
 SummaryGradient/conv5/W/sparsity/Mean: 0
 SummaryGradient/conv6/W-sparsity: 0
 SummaryGradient/conv6/W/sparsity/Mean: 7.4114e-09
 SummaryGradient/conv7/W-sparsity: 0
 SummaryGradient/conv7/W/sparsity/Mean: 1.1209e-08
 SummaryGradient/conv8/W-sparsity: 0
 SummaryGradient/conv8/W/sparsity/Mean: 3.6441e-10
 SummaryGradient/conv8/b-sparsity: 0
 SummaryGradient/conv8/b/sparsity/Mean: 0
 
 gradients/Identity_1_grad/activation-summary/gradients/pool2/output_grad/MaxPoolGrad_0-sparsity: 0.83238
 gradients/Identity_2_grad/activation-summary/gradients/bn3/FusedBatchNorm_grad/FusedBatchNormGrad_0-sparsity: 0
 gradients/Identity_3_grad/activation-summary/gradients/bn4/FusedBatchNorm_grad/FusedBatchNormGrad_0-sparsity: 0
 gradients/Identity_4_grad/activation-summary/gradients/pool3/output_grad/MaxPoolGrad_0-sparsity: 0.84553
 gradients/Identity_5_grad/activation-summary/gradients/bn6/FusedBatchNorm_grad/FusedBatchNormGrad_0-sparsity: 0
 gradients/Identity_6_grad/activation-summary/gradients/bn7/FusedBatchNorm_grad/FusedBatchNormGrad_0-sparsity: 0
 gradients/Identity_grad/activation-summary/gradients/pool1/output_grad/MaxPoolGrad_0-sparsity: 0.8218
```
 
SummaryGradients* are output based on Method 2 and gradients/Identity* are outputs based on method 1.
Here is the application code:  
 
```
 logits = (LinearWrap(image)
              .Conv2D('conv1', 96, 11, stride=4, padding='VALID')
              .apply(fg)
              .MaxPooling('pool1', 3, 2, padding='VALID')
              .BatchNorm('bn1')
              .apply(activate)

              .Conv2D('conv2', 256, 5, padding='SAME', split=2)
              .apply(fg)
              .MaxPooling('pool2', 3, 2, padding='VALID')
              .BatchNorm('bn2')
              .apply(activate)

              .Conv2D('conv3', 384, 3)
              .apply(fg)
              .BatchNorm('bn3')
              .apply(activate)

              .Conv2D('conv4', 384, 3, split=2)
              .apply(fg)
              .BatchNorm('bn4')
              .apply(activate)
              
              .Conv2D('conv5', 256, 3, split=2)
              .apply(fg)
              .MaxPooling('pool3', 3, 2, padding='VALID')
              .BatchNorm('bn5')
              .apply(activate)
          
              .Conv2D('conv6', 4096, 5, padding='VALID')
              .apply(fg)
              .BatchNorm('bn6')
              .apply(activate)
          	  
	      .Conv2D('conv7', 4096, 1)
	      .apply(fg)
              .BatchNorm('bn7')

	      .Conv2D('conv8', 1000, 1, stride=2, use_bias=True, nl=tf.identity)())
```

**Comment by ppwwyyxx** [comment]
Your method summarizes gradients w.r.t layer outputs, not variables.

**Comment by a-maci** [comment]
So, method 2 is correct then?

**Comment by ppwwyyxx** [comment]
It's your decision what to summarize.

**Comment by a-maci** [comment]
Actually I need gradient stats for grads wrt variables(weights) and the grads that get back-propagated to the previous layer. 

Goal is to look at both the grad stats - grads that are involved with weight update and the grad stats that get back-prop'ed to the previous layer (to update weights in that layer). 

Any suggestions?

**Comment by ppwwyyxx** [comment]
Then just summarize both...what's the question?

**Comment by a-maci** [comment]
Ok. Got it.
